dockers/docker-fpm-frr/frr/bgpd/templates/voq_chassis/instance.conf.j2

```diff
@@ -0,0 +1,35 @@
+!
+! template: bgpd/templates/voq_chassis/instance.conf.j2
+!
+  bgp bestpath as-path multipath-relax
+  bgp bestpath peer-type multipath-relax
+!
+  neighbor {{ neighbor_addr }} peer-group VOQ_CHASSIS_PEER
+  neighbor {{ neighbor_addr }} remote-as {{ bgp_session['asn'] }}
+  neighbor {{ neighbor_addr }} description {{ bgp_session['name'] }}
+{# set the bgp neighbor timers if they have not default values #}
+{% if     (bgp_session['keepalive'] is defined and bgp_session['keepalive'] | int != 60)
+      or  (bgp_session['holdtime'] is defined  and bgp_session['holdtime']  | int != 180) %}
+  neighbor {{ neighbor_addr }} timers {{ bgp_session['keepalive'] | default("60") }} {{ bgp_session['holdtime'] | default("180") }}
+{% endif %}
+!
```

Comment:
Can we start using the generic BGP schemas defined to suit any BGP use-cases instead of defining multiple new config-DB schemas per use-case? the below HLD captures all possible BGP configs for a typical DC deployment, this will be complex if we start defining the YANG models for various config-DB schemas defined for different use-cases e.g VOQ chassis.

https://github.com/Azure/SONiC/blob/e311a870a74d311bea1f3ece4a0d7042ed84d0dc/doc/mgmt/SONiC_Design_Doc_Unified_FRR_Mgmt_Interface.md

The backend handling for generic config DB schemas is already merged.
https://github.com/Azure/sonic-buildimage/pull/5142

Reply:
Hi Venkat, when we discussed this at the design review for this feature, we settled on adding a new config DB schema for this (https://github.com/Azure/SONiC/pull/674/files)

The main reason for this was that the voq_chassis peering instances are meant to be 'internal' to the chassis and had specific configurations that won't appear in the other BGP configurations. If we use the generic BGP schema, how would we specify the configs that are only applicable for the voq_chassis instances?

Reply:
>> The main reason for this was that the voq_chassis peering instances are meant to be 'internal' to the chassis and had specific configurations that won't appear in the other BGP configurations.

I assume VoQ_ chassis configs will be present in config-DB and will be converted to FRR configs with the help of template files, when you say 'internal' what exactly the behaviour and how it wont appear in the other BGP configurations, I believe, FRR treats all BGP configurations including internal or external as same, I'm trying to understand the importance of defining new schemas, does it help to filter anyways? or it can be generic as well.

Reply:
Yes, the voq_chassis configs will be present in Config DB and will be converted to FRR configs with the help of the template files. 

By 'internal', I mean that they won't be displayed in some of the Bgp related show commands or be affected by some of the bgp config commands. From FRR point of view, there is no difference, correct. Some FRR configs like "bgp bestpath as-path multipath-relax" is specific to the voq_chassis peers, for example. And we don't expect this config to be changed by the operator an runtime. 

If I understood you right, you are proposing to update the generic table to map to these new commands and have the voq_chassis specific flags/configs defined in the Sonic DB schema? I suppose that could be done, do you have an example of this? We spent quite a bit of time debating this at the design review and followed the existing model where each BGP peer type defined a new template.

Reply:
I did some of the VoQ BGP configs using generic BGP models and observed the following, see if this helps for VoQ.

Note: In order for the generic BGP model configs to handle using "src/sonic-frr-mgmt-framework/frrcfgd", we should set the field  "frr_mgmt_framework_config" to true in DEVICE_METADATA table, this was introduced to keep config handling by bgpcfgd as default in the community.

Also, config restore is supported, we read configs from config-DB and push to FRR during restore.

```

admin@sonic:~$ redis-cli -n 4
127.0.0.1:6379[4]> keys BGP*
1) "BGP_PEER_GROUP_AF|default|VOQ_CHASSIS_PEER|ipv6_unicast"
2) "BGP_GLOBALS_AF|default|ipv4_unicast"
3) "BGP_NEIGHBOR|default|10.10.10.10"
4) "BGP_GLOBALS_AF|default|ipv6_unicast"
5) "BGP_GLOBALS|default"
6) "BGP_PEER_GROUP|default|VOQ_CHASSIS_PEER"
7) "BGP_PEER_GROUP_AF|default|VOQ_CHASSIS_PEER|ipv4_unicast"
127.0.0.1:6379[4]> hgetall BGP_GLOBALS|default
 1) "as_path_mp_as_set"
 2) "false"
 3) "local_asn"
 4) "65100"
 5) "log_nbr_state_changes"
 6) "true"
 7) "always_compare_med"
 8) "false"
 9) "external_compare_router_id"
10) "false"
11) "fast_external_failover"
12) "true"
13) "network_import_check"
14) "true"
15) "load_balance_mp_relax"
16) "true"
17) "holdtime"
18) "30"
19) "ignore_as_path_length"
20) "false"
21) "keepalive"
22) "5"
127.0.0.1:6379[4]> hgetall BGP_GLOBALS_AF|default|ipv4_unicast
1) "route_flap_dampen"
2) "false"
3) "ibgp_equal_cluster_length"
4) "false"
5) "max_ebgp_paths"
6) "1"
7) "max_ibgp_paths"
8) "64"
127.0.0.1:6379[4]> hgetall BGP_GLOBALS_AF|default|ipv6_unicast
1) "ibgp_equal_cluster_length"
2) "false"
3) "max_ebgp_paths"
4) "1"
5) "max_ibgp_paths"
6) "64"
127.0.0.1:6379[4]> hgetall BGP_NEIGHBOR|default|10.10.10.10
1) "holdtime"
2) "240"
3) "peer_group_name"
4) "VOQ_CHASSIS_PEER"
5) "name"
6) "internal1"
7) "keepalive"
8) "60"
127.0.0.1:6379[4]> hgetall BGP_PEER_GROUP|default|VOQ_CHASSIS_PEER
 1) "min_adv_interval"
 2) "0"
 3) "ebgp_multihop"
 4) "false"
 5) "passive_mode"
 6) "false"
 7) "admin_status"
 8) "true"
 9) "conn_retry"
10) "30"
11) "asn"
127.0.0.1:6379[4]> hgetall BGP_PEER_GROUP_AF|default|VOQ_CHASSIS_PEER|ipv4_unicast
 1) "rrclient"
 2) "false"
 3) "admin_status"
 4) "true"
 5) "send_default_route"
 6) "false"
 7) "send_community"
 8) "both"
 9) "max_prefix_warning_only"
10) "false"
11) "NULL"
12) "NULL"
127.0.0.1:6379[4]>
127.0.0.1:6379[4]> hgetall BGP_PEER_GROUP_AF|default|VOQ_CHASSIS_PEER|ipv6_unicast
 1) "NULL"
 2) "NULL"
 3) "admin_status"
 4) "true"
 5) "rrclient"
 6) "false"
 7) "send_default_route"
 8) "false"
 9) "max_prefix_warning_only"
10) "false"
11) "send_community"
12) "both"
127.0.0.1:6379[4]>


admin@sonic:~$ vtysh
Hello, this is FRRouting (version 7.2-sonic).
Copyright 1996-2005 Kunihiro Ishiguro, et al.

sonic#
sonic#
sonic# show running-config bgpd
Building configuration...

Current configuration:
!
frr version 7.2-sonic
frr defaults traditional
hostname sonic
log syslog informational
log facility local4
agentx
no service integrated-vtysh-config
!
enable password zebra
password zebra
!
router bgp 65100
 bgp log-neighbor-changes
 no bgp default ipv4-unicast
 bgp bestpath as-path multipath-relax
 bgp bestpath compare-routerid
 bgp network import-check
 timers bgp 5 30
 neighbor VOQ_CHASSIS_PEER peer-group
 neighbor VOQ_CHASSIS_PEER remote-as 555
 neighbor VOQ_CHASSIS_PEER advertisement-interval 0
 neighbor VOQ_CHASSIS_PEER timers connect 30
 neighbor 10.10.10.10 peer-group VOQ_CHASSIS_PEER
 neighbor 10.10.10.10 description internal1
 neighbor 10.10.10.10 timers 60 240
 !
 address-family ipv4 unicast
  neighbor VOQ_CHASSIS_PEER activate
  no neighbor VOQ_CHASSIS_PEER send-community large
  no neighbor 10.10.10.10 send-community large
  maximum-paths 1
  maximum-paths ibgp 64
 exit-address-family
 !
 address-family ipv6 unicast
  neighbor VOQ_CHASSIS_PEER activate
  no neighbor VOQ_CHASSIS_PEER send-community large
  maximum-paths 1
  maximum-paths ibgp 64
 exit-address-family
!
line vty
!
end
sonic# show bgp ipv4 unicast summary
BGP router identifier 100.94.134.22, local AS number 65100 vrf-id 0
BGP table version 0
RIB entries 0, using 0 bytes of memory
Peers 1, using 20 KiB of memory
Peer groups 1, using 64 bytes of memory

Neighbor        V         AS MsgRcvd MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd
10.10.10.10     4        555       0       0        0    0    0    never       Active

Total number of neighbors 1
Total number of neighbors established 0
sonic# show bgp ipv6 unicast summary
BGP router identifier 100.94.134.22, local AS number 65100 vrf-id 0
BGP table version 0
RIB entries 0, using 0 bytes of memory
Peers 1, using 20 KiB of memory
Peer groups 1, using 64 bytes of memory

Neighbor        V         AS MsgRcvd MsgSent   TblVer  InQ OutQ  Up/Down State/PfxRcd
10.10.10.10     4        555       0       0        0    0    0    never       Active

Total number of neighbors 1
Total number of neighbors established 0
sonic#


```

Reply:
@venkatmahalingam, this feature is designed to work with bgpcfgd, not frrcfgd.

Reply:
Okay, no problem.

Reply:
Aah, I did not realize the distinction earlier. @venkatmahalingam please let us know if there is any change required here.

Reply:
@eschwabe There are two approaches we follow in the community for FRR configurations 1) bgpcfgd 2) frrcfgd, @lguohan mentioned that VoQ chassis is designed to work with bgpcfgd, so, we can go ahead with that, changing to adapt to frrcfgd might be a lot of work at this point.